[FLINK-32201][runtime]Automatically determine if the shuffle descriptor needs to be offloaded by the blob server based on the number of ShuffleDescriptor edges.

huwh opened the pull request:
## What is the purpose of the change

Flink support distributes shuffle descriptors via the blob server to reduce JobManager overhead. But the default threshold to enable it is 1MB, which never reaches. Users need to set a proper value for this, but it requires advanced knowledge before configuring it.

I would like to enable this feature by the number of connections of a group of shuffle descriptors. For examples, a simple streaming job with two operators, each with 10,000 parallelism and connected via all-to-all distribution. In this job, we only get one set of shuffle descriptors, and this group has 10000 * 10000 connections. This means that JobManager needs to send this set of shuffle descriptors to 10000 tasks.

Since it's difficult for users to configure, I would like to give it a default value. 

## Brief change log
  - *determine if the shuffle descriptor needs to be offloaded by the blob server based on the number of ShuffleDescriptor edges.*

## Verifying this change

This change added tests and can be verified as follows:
  - *Manually verified the change by running a cluster with 1 JobManagers and 2000 TaskManagers (10 slots per TaskManager), a streaming program with 20000 parallelism, and verifying that the task deploy was successful.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  (no)
  - The serializers:  (no)
  - The runtime per-record code paths (performance sensitive):  (no)
  - Anything that affects deployment or recovery: Task Deployment: (yes)
  - The S3 file system connector:  (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -104,4 +122,14 @@ public void markPartitionFinished(IntermediateResultPartition resultPartition) {
+
+    private boolean isForceOffload(ShuffleDescriptorAndIndex[] shuffleDescriptorsToSerialize) {
+        // The unknown shuffle descriptor will be compressed, so don't take it into account.
+        long numKnownShuffleDescriptors =
```
Could we add some tests for the new logics, particularly for this part?

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -104,4 +122,14 @@ public void markPartitionFinished(IntermediateResultPartition resultPartition) {
                                 resultPartitionIdToIndex.get(resultPartition.getPartitionId()))));
     }
+
+    private boolean isForceOffload(ShuffleDescriptorAndIndex[] shuffleDescriptorsToSerialize) {
```
How about `shouldOffload` instead of `isForceOffload`? I think we're not forced to do this optimization.🤣

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -38,6 +39,14 @@
+     * to serialized and transport 10000 shuffle descriptors(almost 200KB) to 10000 consumer(2GB in
+     * total)
+     */
+    private static final int FORCE_OFFLOAD_SHUFFLE_DESCRIPTORS_THRESHOLD = 10000 * 10000;
```
How about making this option configurable but not exposed to the user manual?

Thesharing reviewed (commented):
Thank you for creating this pull request, @huwh!👍 I left several comments here. Please correct me if anything I said is wrong.

huwh commented:
Thanks for your review. @Thesharing , your comments are totally correct. 

I updated the PR. Please take a look at your free time.

There is a change beyond the comments. I removed the logic to exclude Unknown shuffle descriptors, as it also made the serialized shuffle descriptors larger (200KB for 10000).

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -54,10 +56,18 @@ public class CachedShuffleDescriptors {
+    /** The number of consumers for {@link ConsumedPartitionGroup}. */
+    private final int numConsumers;
+
+    private final int offloadShuffleDescriptorsThreshold;
```
Since all the values of `offloadShuffleDescriptorsThreshold` are the same, could we put this value into `TaskDeploymentDescriptorFactory`? Also, we could move the `shouldOffload` function to the `TaskDeploymentDescriptorFactory`, too. Then we just need to pass all necessary parameters to `shouldOffload` in `serializeAndTryOffloadShuffleDescriptor` function.

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -216,14 +221,16 @@ private ShuffleDescriptorAndIndex[] computeConsumedPartitionShuffleDescriptors(
         final Either<SerializedValue<ShuffleDescriptorAndIndex[]>, PermanentBlobKey>
                 serializedValueOrBlobKey =
-                        BlobWriter.tryOffload(compressedSerializedValue, jobID, blobWriter);
+                        BlobWriter.tryOffload(
```
Since `BlobWriter` don't know the logic of task deployment, here we could say:

```java
final Either<SerializedValue<ShuffleDescriptorAndIndex[]>, PermanentBlobKey>
        serializedValueOrBlobKey =
                shouldOffload(numConsumers)
                        ? BlobWriter.offloadWithException(
                                compressedSerializedValue, jobID, blobWriter)
                        : Either.Left(compressedSerializedValue);
```

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -104,4 +119,30 @@ public void markPartitionFinished(IntermediateResultPartition resultPartition) {
     }
+
+    private boolean shouldOffload(ShuffleDescriptorAndIndex[] shuffleDescriptorsToSerialize) {
+        return shuffleDescriptorsToSerialize.length * numConsumers
```
I'm wondering why we use `shuffleDescriptorsToSerialize.length` here. It may be a bit confusing for this case:

For a vertex of 10,000 * 10,000 in hybrid shuffle:
1. `shuffleDescriptorsToSerialize` only contains half of the ShuffleDescriptors, it wouldn't be offloaded.
2. Afterwards, `shuffleDescriptorsToSerialize` contains all the ShuffleDescriptors, it would be offloaded.

Could we just use `numConsumers` here?

huwh replied:
In some scenarios, there may only be a few new shuffleDescriptors to serialize, so it is more efficient to use Akka to transfer them directly.

Thesharing commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphBuilder.java:
```diff
@@ -143,7 +147,8 @@ public static DefaultExecutionGraph buildGraph(
                             jobGraph.getJobStatusHooks(),
                             markPartitionFinishedStrategy,
-                            nonFinishedHybridPartitionShouldBeUnknown);
+                            nonFinishedHybridPartitionShouldBeUnknown,
```
Hey, @reswqa. I'm wondering since `nonFinishedHybridPartitionShouldBeUnknown` and `offloadShuffleDescriptorsThreshold` are both related to task deployment, could we merge them into a configuration class or a factory or something else? It could make the initialization of `ExecutionGraph` and other related components more readable. What's your opinion?

reswqa replied:
Thanks for inviting me. I didn't carefully check the code, but I think they can be merged into one auxiliary class. If we plan to do this refactoring, other task deployment related fields (if any) should also be merged in.

huwh replied:
Thanks @Thesharing @reswqa. I created a new ticket (https://issues.apache.org/jira/browse/FLINK-32225) to do merge-related work to make this PR clean. I will finish that ticket first.

Thesharing reviewed (commented):
Thank you for the update, @huwh! 👍 I've left some comments here. Please correct me if anything I said is wrong.

huwh commented:
@Thesharing, @wanglijie95 , I have rebase this to the master branch. Could you please take another look?

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -221,14 +224,17 @@ private ShuffleDescriptorAndIndex[] computeConsumedPartitionShuffleDescriptors(
 
     private MaybeOffloaded<ShuffleDescriptorAndIndex[]> serializeAndTryOffloadShuffleDescriptor(
-            ShuffleDescriptorAndIndex[] shuffleDescriptors) throws IOException {
+            ShuffleDescriptorAndIndex[] shuffleDescriptors, int numConsumer) throws IOException {
```
Can we abstract the logic here to a inline class `ShuffleDescriptorSerializer`, and provide 2 method: `serializeShuffleDescriptor` and `serializeAndTryOffloadShuffleDescriptor`. And then pass the serializer to `CachedShuffleDescriptors`, the `CachedShuffleDescriptors` will decide to call which method according to the `shouldOffload`

wanglijie95 reviewed (changes requested):
Thanks for creating this PR @huwh . I left some comments, PTAL.

wanglijie95 commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -717,6 +717,18 @@ public boolean isOnlyConsumeFinishedPartition() {
 
+    @Documentation.ExcludeFromDocumentation(
+            "this is an expert option, that we do not want to expose in the documentation")
+    public static final ConfigOption<Integer> OFFLOAD_SHUFFLE_DESCRIPTORS_THRESHOLD =
```
I suggest to move this config option to `TaskDeploymentDescriptorFactory` and marked as `@Experimental`. We can put it back to `JobManagerOptions`  when we think it should be exposed to users

huwh commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/CachedShuffleDescriptors.java:
```diff
@@ -54,10 +54,14 @@ public class CachedShuffleDescriptors {
+
     public CachedShuffleDescriptors(
             ConsumedPartitionGroup consumedPartitionGroup,
             ShuffleDescriptorAndIndex[] shuffleDescriptors) {
```
There could be some incremental serialization of ShuffleDescriptors, and may cause performance regression if there are few shuffle descriptors

huwh commented:
@wanglijie95 Thanks for your review, I have updated this PR, PTAL

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -428,33 +434,66 @@ public int getIndex() {
     }
+
+    @Experimental
+    @Documentation.ExcludeFromDocumentation(
```
I think the `@Documentation.ExcludeFromDocumentation` is not needed here. We can change it to comments.  And we generally put the option at the front of a class.

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -428,33 +434,66 @@ public int getIndex() {
+                    .withDescription(
+                            "The threshold to force enable offload shuffle descriptors via blob server."
+                                    + " This is a fixed value since it is difficult for users to configure."
+                                    + " This default value means JobManager need to serialized and transport"
```
`This is a fixed value since it is difficult for users to configure` -> This comments is outdated.

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -428,33 +434,66 @@ public int getIndex() {
     interface ShuffleDescriptorSerializer {
         MaybeOffloaded<ShuffleDescriptorAndIndex[]> serializeAndTryOffloadShuffleDescriptor(
-                ShuffleDescriptorAndIndex[] shuffleDescriptors) throws IOException;
+                ShuffleDescriptorAndIndex[] shuffleDescriptors, int numConsumer) throws IOException;
```
We'd better add document for this method, especially the meaning of params.

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -428,33 +434,66 @@ public int getIndex() {
+        public DefaultShuffleDescriptorSerializer(
+                JobID jobID, BlobWriter blobWriter, int offloadShuffleDescriptorsThreshold) {
             this.jobID = jobID;
             this.blobWriter = blobWriter;
```
Add `checkNotNull` for `jobID` and `blobWriter`

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -58,30 +60,45 @@
 public class TaskDeploymentDescriptorFactory {
+    @Experimental
+    public static final ConfigOption<Integer> OFFLOAD_SHUFFLE_DESCRIPTORS_THRESHOLD =
+            key("jobmanager.deployment.offload-shuffle-descriptors-threshold")
```
Maye be change the key to `jobmanager.task-deployment.offload-shuffle-descriptors-to-blob-server.threshold-num` ?

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -58,30 +60,45 @@
+                    .intType()
+                    .defaultValue(4_000 * 4_000)
+                    .withDescription(
+                            "The threshold to force enable offload shuffle descriptors via blob server."
```
Maybe change the first sentence to "Threshold for offloading shuffle descriptors to blob server. Once the number of shuffle descriptors exceeds this value, we will offload the shuffle descriptors to blob server."

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -58,30 +60,45 @@
+                    .withDescription(
+                            "The threshold to force enable offload shuffle descriptors via blob server."
+                                    + " This default value means JobManager need to serialized and transport"
+                                    + " 4000 shuffle descriptors(almost 100KB) to 4000 consumer(400MB in total)");
```
```suggestion
                                    + " 4000 shuffle descriptors(almost 100KB) to 4000 consumers (400MB in total)");
```

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -58,30 +60,45 @@
+                    .defaultValue(4_000 * 4_000)
+                    .withDescription(
+                            "The threshold to force enable offload shuffle descriptors via blob server."
+                                    + " This default value means JobManager need to serialized and transport"
```
```suggestion
                                    + " This default value means JobManager need to serialize and transport"
```

wanglijie95 commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java:
```diff
@@ -58,30 +60,45 @@
  */
 public class TaskDeploymentDescriptorFactory {
+    @Experimental
+    public static final ConfigOption<Integer> OFFLOAD_SHUFFLE_DESCRIPTORS_THRESHOLD =
```
Maybe add doc for this option, and explain it.  For example, "This is an expert option, that we do not want to expose in the documentation. The default value is good enough for almost all cases"

wanglijie95 reviewed (commented):
@huwh I left some minor comments about the configuration option, PTAL.

huwh commented:
@wanglijie95 ，Thanks for pointing out the problems. I have addressed them as you suggested.

huwh commented:
We give this option a default value  2048 * 2048, which is good enough for most cases. 
This means JobManager need to serialize and transport 2048 shuffle descriptors (about 30KB) to 2048 consumers (60MB in total).

wanglijie95 commented:
@flinkbot run azure
